src/include/alloc.h

```diff
@@ -106,19 +108,70 @@ niova_vbasic_nassigned(const struct niova_vbasic_allocator *nvba)
 }
 
 static inline int
-niova_vbasic_init(struct niova_vbasic_allocator *nvba,
-                  size_t region_size)
+niova_vbasic_init(struct niova_vbasic_allocator *nvba, size_t region_size)
 {
     if (!nvba)
         return -EINVAL;
 
     unsigned int unit_size = region_size / NIOVA_VBA_MAX_BITS;
+
     if (unit_size == 0)
         return -EINVAL;
 
     CONST_OVERRIDE(size_t, nvba->nvba_unit_size, unit_size);
     nvba->nvba_bitmap = 0;
+    nvba->nvba_alignment = 0;
+
+    nvba->nvba_region_ptr = &nvba->nvba_region[0];
+    return 0;
+}
+
+/**
+ * To callers: Always allocate region size inclusive of alignment
+ */
+static inline int
+niova_vbasic_init_aligned(struct niova_vbasic_allocator *nvba,
+                          size_t region_size, unsigned int unit_size,
+                          size_t alignment)
+{
+    if (!nvba)
+        return -EINVAL;
+
+    if (unit_size == 0 || region_size == 0)
+        return -EINVAL;
+
+    /* alignment is power of 2 */
+    if (alignment == 0 || (alignment & (alignment - 1)))
+        return -EINVAL;
+
+    if (alignment > unit_size)
+        return -EINVAL;
+
+    /* Unit size is multiple of alignment */
+    if (unit_size & (alignment - 1))
+        return -EINVAL;
+
+    if (region_size < NIOVA_VBA_MAX_BITS)
+        return -EINVAL;
+
+    if (region_size / unit_size > NIOVA_VBA_MAX_BITS)
+        return -EINVAL;
+
+    const unsigned int nunits =
+        (region_size / unit_size) + (region_size % unit_size ? 1 : 0);
+
+    if (nunits > NIOVA_VBA_MAX_BITS)
+        return -EINVAL;
+
+    CONST_OVERRIDE(size_t, nvba->nvba_unit_size, unit_size);
+    nvba->nvba_bitmap = 0;
```

Comment:
Should we deal with cases where the number of units is < 64? If so, we'd need to mark the bits of the unallocatable regions as '1'.